layers/+lang/haskell/funcs.el

```diff
@@ -22,15 +22,98 @@
     (haskell-navigate-imports)
     (haskell-mode-format-imports)))
 
-;; Dante Functions
+;; Completion setup functions
+
+(defun spacemacs-haskell//setup-backend ()
+  "Conditionally setup haskell backend."
+  (pcase haskell-completion-backend
+    (`ghci (spacemacs-haskell//setup-ghci))
+    (`lsp (spacemacs-haskell//setup-lsp))
+    (`intero (spacemacs-haskell//setup-intero))
+    (`dante (spacemacs-haskell//setup-dante))
+    (`ghc-mod (spacemacs-haskell//setup-ghc-mod))))
+
+(defun spacemacs-haskell//setup-company ()
+  "Conditionally setup haskell backend."
+  (pcase haskell-completion-backend
+    (`ghci (spacemacs-haskell//setup-ghci-company))
+    (`lsp (spacemacs-haskell//setup-lsp-company))
+    (`intero (spacemacs-haskell//setup-intero-company))
+    (`dante (spacemacs-haskell//setup-dante-company))
+    (`ghc-mod (spacemacs-haskell//setup-ghc-mod-company))))
+
+;; ghci functions
+
+(defun spacemacs-haskell//setup-ghci ()
+  (interactive-haskell-mode))
+
+(defun spacemacs-haskell//setup-ghci-company ()
+  (spacemacs|add-company-backends
+    :backends (company-ghci company-dabbrev-code company-yasnippet)
+    :modes haskell-mode))
+
+;; LSP functions
+
+(defun spacemacs-haskell//setup-lsp ()
+  "Setup lsp backend"
+  (if (configuration-layer/layer-used-p 'lsp)
+      (progn
+        ;; The functionality we require from this is not an autoload, but rather some
+        ;; top-level code that registers a LSP server type. So we need to load it
+        ;; directly and can't rely on it being autoloaded.
+        (require 'lsp-haskell)
+        (lsp))
+    (message "`lsp' layer is not installed, please add `lsp' layer to your dotfile.")))
+
+(defun spacemacs-haskell//setup-lsp-company ()
```

Comment:
no longer needed in latest lsp-mode, it will automatically configure lsp-mode

Reply:
It looks like all the other layers still do this - are we sure to get the newest lsp-mode?

Reply:
These are pretty much leftovers, there are PRs for fixing some of them (like https://github.com/syl20bnr/spacemacs/pulls/DanielRichtmann ). The version that introduced autoconfiguration is the same as the version that introduced `lsp` function.

Reply:
Okay, I'll try and update this to do the right thing then :+1:

Reply:
Done, seems to work!